SYCL/ESIMD/PrefixSum.cpp

```diff
@@ -0,0 +1,421 @@
+              << " thread_stride: " << thrd_stride
+              << " entry per thread: " << entry_per_th << std::endl;
+    // one single thread
+    q.submit([&](handler &cgh) {
+      cgh.parallel_for<class Accum_final>(
+          range<2>{1, 1} * range<2>{1, 1}, [=](item<2> it) SYCL_ESIMD_KERNEL {
+            cmk_acum_final(buf, it.get_id(0), elem_stride, n_entries);
+          });
+    });
+    return;
+  }
+
+  std::cout << "*** n_entries: " << n_entries << " elem_stide: " << elem_stride
+            << " thread_stride: " << thrd_stride
+            << " entry per thread: " << entry_per_th << std::endl;
+
+  if (entry_per_th == PREFIX_ENTRIES) {
+    q.submit([&](handler &cgh) {
+      cgh.parallel_for<class Accum_iterative1>(
+          range<2>{n_entries / entry_per_th, 1} * range<2>{1, 1},
+          [=](item<2> it) SYCL_ESIMD_KERNEL {
+            cmk_acum_iterative(buf, it.get_id(0), elem_stride, thrd_stride,
+                               PREFIX_ENTRIES);
+          });
+    });
+  } else {
+    q.submit([&](handler &cgh) {
+      cgh.parallel_for<class Accum_iterative2>(
+          range<2>{n_entries / entry_per_th, 1} * range<2>{1, 1},
+          [=](item<2> it) SYCL_ESIMD_KERNEL {
+            cmk_acum_iterative(buf, it.get_id(0), elem_stride, thrd_stride,
+                               PREFIX_ENTRIES_LOW);
+          });
+    });
+  }
+
+  // if number of remaining entries <= 4K , each thread  accumulates smaller
+  // number of entries to keep EUs saturated
+  if (n_entries / entry_per_th > 4096)
+    hierarchical_prefix(q, buf, thrd_stride, thrd_stride * PREFIX_ENTRIES,
+                        n_entries / entry_per_th, PREFIX_ENTRIES);
+  else
+    hierarchical_prefix(q, buf, thrd_stride, thrd_stride * PREFIX_ENTRIES_LOW,
+                        n_entries / entry_per_th, PREFIX_ENTRIES_LOW);
+
+  std::cout << "=== n_entries: " << n_entries << " elem_stide: " << elem_stride
+            << " thread_stride: " << thrd_stride
+            << " entry per thread: " << entry_per_th << std::endl;
+
+  if (entry_per_th == PREFIX_ENTRIES) {
+    q.submit([&](handler &cgh) {
+      cgh.parallel_for<class Prefix_iterative1>(
+          range<2>{n_entries / entry_per_th, 1} * range<2>{1, 1},
+          [=](item<2> it) SYCL_ESIMD_KERNEL {
+            cmk_prefix_iterative(buf, it.get_id(0), elem_stride, thrd_stride,
+                                 PREFIX_ENTRIES);
+          });
+    });
+  } else {
+    q.submit([&](handler &cgh) {
+      cgh.parallel_for<class Prefix_iterative2>(
+          range<2>{n_entries / entry_per_th, 1} * range<2>{1, 1},
+          [=](item<2> it) SYCL_ESIMD_KERNEL {
+            cmk_prefix_iterative(buf, it.get_id(0), elem_stride, thrd_stride,
+                                 PREFIX_ENTRIES_LOW);
+          });
+    });
+  }
+}
+
+//************************************
+// Demonstrate summation of arrays both in scalar on CPU and parallel on device
+//************************************
```

Comment:
I suggest to add some differentiating comments into the beginning of each test, to give an idea what is the difference between them.

Reply:
ESIMD has only a few numbers of tests. As we developed PrefixSum, there were some intermediate steps that didn’t end up in the final implementation, but we like to turn them into ULT to increase our coverage of different usage cases.  They are smaller tests which allow compiler to easily isolate bugs.

We have added description to the ULT tests to summarize the difference in the new commit.